packages/core/src/client.ts

```diff
@@ -0,0 +1,179 @@
+	[GatewayDispatchEvents.MessageReactionRemoveEmoji]: [
+		WithIntrinsicProps<GatewayMessageReactionRemoveEmojiDispatchData>,
+	];
+	[GatewayDispatchEvents.MessageUpdate]: [WithIntrinsicProps<GatewayMessageUpdateDispatchData>];
+	[GatewayDispatchEvents.PresenceUpdate]: [WithIntrinsicProps<GatewayPresenceUpdateDispatchData>];
+	[GatewayDispatchEvents.Ready]: [WithIntrinsicProps<GatewayReadyDispatchData>];
+	[GatewayDispatchEvents.StageInstanceCreate]: [WithIntrinsicProps<GatewayStageInstanceCreateDispatchData>];
+	[GatewayDispatchEvents.StageInstanceDelete]: [WithIntrinsicProps<GatewayStageInstanceDeleteDispatchData>];
+	[GatewayDispatchEvents.StageInstanceUpdate]: [WithIntrinsicProps<GatewayStageInstanceUpdateDispatchData>];
+	[GatewayDispatchEvents.ThreadCreate]: [WithIntrinsicProps<GatewayThreadCreateDispatchData>];
+	[GatewayDispatchEvents.ThreadDelete]: [WithIntrinsicProps<GatewayThreadDeleteDispatchData>];
+	[GatewayDispatchEvents.ThreadListSync]: [WithIntrinsicProps<GatewayThreadListSync>];
+	[GatewayDispatchEvents.ThreadMemberUpdate]: [WithIntrinsicProps<GatewayThreadMemberUpdateDispatchData>];
+	[GatewayDispatchEvents.ThreadMembersUpdate]: [WithIntrinsicProps<GatewayThreadMembersUpdateDispatchData>];
+	[GatewayDispatchEvents.ThreadUpdate]: [WithIntrinsicProps<GatewayThreadUpdateDispatchData>];
+	[GatewayDispatchEvents.UserUpdate]: [WithIntrinsicProps<GatewayUserUpdateDispatchData>];
+	[GatewayDispatchEvents.VoiceServerUpdate]: [WithIntrinsicProps<GatewayVoiceServerUpdateDispatchData>];
+	[GatewayDispatchEvents.VoiceStateUpdate]: [WithIntrinsicProps<GatewayVoiceStateUpdateDispatchData>];
+	[GatewayDispatchEvents.WebhooksUpdate]: [WithIntrinsicProps<GatewayWebhooksUpdateDispatchData>];
+	[GatewayDispatchEvents.Resumed]: [WithIntrinsicProps<GatewayResumedDispatch['d']>];
+	[GatewayDispatchEvents.TypingStart]: [WithIntrinsicProps<GatewayTypingStartDispatchData>];
+	[GatewayDispatchEvents.AutoModerationActionExecution]: [
+		WithIntrinsicProps<GatewayAutoModerationActionExecutionDispatchData>,
+	];
+	[GatewayDispatchEvents.AutoModerationRuleCreate]: [WithIntrinsicProps<GatewayAutoModerationRuleCreateDispatchData>];
+	[GatewayDispatchEvents.AutoModerationRuleDelete]: [WithIntrinsicProps<GatewayAutoModerationRuleDeleteDispatchData>];
+	[GatewayDispatchEvents.AutoModerationRuleUpdate]: [WithIntrinsicProps<GatewayAutoModerationRuleUpdateDispatchData>];
+}
+
+export type ManagerShardEventsMap = {
+	[K in keyof MappedEvents]: MappedEvents[K];
+};
+
+export interface ClientOptions {
+	rest: REST;
+	ws: WebSocketManager;
+}
+
+export function createClient({ rest, ws }: ClientOptions) {
+	const eventAPI = new API(rest);
+	const emitter = new AsyncEventEmitter<ManagerShardEventsMap>();
+
+	ws.on(WebSocketShardEvents.Dispatch, ({ data: dispatch, shardId }) => {
+		function wrapIntrinsicProps<T>(obj: T): WithIntrinsicProps<T> {
+			return {
+				api: eventAPI,
+				shardId,
+				data: obj,
+			};
+		}
```

Comment:
This can be inefficient, as it'll create a function (which has some overhead) on every dispatch.

Instead, you can move it outside of the listener (can be inside `createClient` to read `eventAPI`) and take `shardId` as a parameter.